File: snippets/0001_two_sum.py
from typing import List


# Hashmap
def twoSum(nums: List[int], target: int) -> List[int]:
    hashmap = dict()  # {val: idx}

    for idx, val in enumerate(nums):
        if (target - val) in hashmap:
            return [hashmap[target - val], idx]
        hashmap[val] = idx


# |-------------|-----------------|--------------|
# |  Approach   |      Time       |    Space     |
# |-------------|-----------------|--------------|
# |  Hashmap    |      O(n)       |    O(n)      |
# |-------------|-----------------|--------------|


nums = [2, 7, 11, 15]
target = 18
print(twoSum(nums, target))  # [1, 2]

File: snippets/js/0001_two_sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
  let map = new Map(); // {val: idx}

  for (let i = 0; i < nums.length; i++) {
    let complement = target - nums[i];

    if (map.has(complement)) {
      return [map.get(complement), i];
    }

    map.set(nums[i], i);
  }
};

const nums = [2, 7, 11, 15];
const target = 9;
console.log(twoSum(nums, target));
